### PR TITLE
Illegal character underscore in taskname and Acq name

### LINKED ISSUE #221: Check for underscores in task names (and possibly other labels)
Currently BIDS does not allow underscores to be used in labels. For example:

`sub-269/func/sub-269_task-reward_rest_bold.nii.gz`

is incorrect (should be: `sub-269/func/sub-269_task-rewardrest_bold.nii.gz`). Currently validator will give a warning that this file does not belong to BIDS. It would be more user friendly to try to detect those cases and give a separate warning about underscores in labels.

### PR DESCRIPTION
**Pull request by suyashdb**
Closes #221

### CODE REVIEW

**Comment by chrisgorgo** [comment]
Is this ready for review? Would it be hard to expand to `sub-` and `ses-` (I just run into another dataset that made such mistake).

**Comment by suyashdb** [comment]
I haven't been able to write correct tests for this PR yet. Is it ok to merge for now without tests?  

Can i create another PR for finding illegal character in sub-/ses-? I think it should be not difficult as finding illegal character in taskname and acq.

**Comment by chrisgorgo** [comment]
Let's wait for tests until we merge this one and do sub-/ses- in another PR.

**Comment by chrisgorgo** [comment]
Try following the pattern of this test: https://github.com/suyashdb/bids-validator/blob/4632dabfa7616dc58398af92915239f4bd820d7f/tests/bids.spec.js#L83

**Comment by suyashdb** [comment]
@chrisfilo Ready for review and merge

File: utils/issues/list.js [hunk 1]
```diff
@@ -283,8 +283,13 @@ module.exports = {
         reason: 'As per section 164.514(C) of "The De-dentification Standard" under HIPPA guidelines, participants with age 89 or higher should be tagged as 89+. More information can be found at https://www.hhs.gov/hipaa/for-professionals/privacy/special-topics/de-identification/#standard'
     },
     57: {
-        key: 'DATASET_DESCRIPTION_JSON_MISSING',
+        key: 'Task Name contain illegal Character -/_',
```
Comment (chrisgorgo) [inline utils/issues/list.js]: let's keep keys as all caps one string (but with underscores)

File: tests/bids.spec.js [hunk 2]
```diff
@@ -90,9 +90,20 @@ var suite = describe('BIDS example datasets ', function() {
...
             isdone();
         });
     });
+
+    // test for illegal characters used in acq and task name
+    it('validates dataset with illegal characters in task name', function(isdone) {
+        var options = {ignoreNiftiHeaders: false};
+        validate.BIDS("tests/data/valid_filenames", options, function (issues) {
+            var errors = issues.errors;
+            assert(errors[0].code === '55' && errors[1].code === '57');
```
Comment (chrisgorgo) [inline tests/bids.spec.js]: let's limit this to one error (57)

File: validators/bids.js [hunk 3]
```diff
@@ -97,38 +98,67 @@ var BIDS = {
...
+            if (acq_res){
+                self.issues.push(new Issue({
+                    file: fileList[fi],
+                    code: 58,
+                    evidence: "acq name contains illegal character: " + fileList[fi].relativePath
+                }));
+            }
+
+        }
+
```
Comment (chrisgorgo) [inline validators/bids.js]: Could you rewrite this so it would be easier to add this check for new keys in the future.

**Review by chrisgorgo — CHANGES_REQUESTED** [review]

**Comment by suyashdb** [comment]
@chrisfilo can you please review and merge?

File: utils/issues/list.js [hunk 4]
```diff
@@ -283,8 +283,13 @@ module.exports = {
         reason: 'As per section 164.514(C) of "The De-dentification Standard" under HIPPA guidelines, participants with age 89 or higher should be tagged as 89+. More information can be found at https://www.hhs.gov/hipaa/for-professionals/privacy/special-topics/de-identification/#standard'
     },
     57: {
-        key: 'DATASET_DESCRIPTION_JSON_MISSING',
+        key: 'TASK_NAME_CONTAIN_ILLEGAL_CHARACTER',
         severity: 'error',
-        reason: 'The compulsory file /dataset_description.json is missing. See Section 8.1 of the BIDS specification.'
+        reason: 'Task Name contain an Illegal Character hypen or underscore. Please edit the filename as per BIDS spec.'
```
Comment (chrisgorgo) [inline utils/issues/list.js]: Why are you changing the name and meaning of this error instead of adding a new one?

File: validators/bids.js [hunk 5]
```diff
@@ -97,38 +98,63 @@ var BIDS = {
...
             subjects: [],
-            tasks:    [],
+            tasks: [],
             modalities: [],
             totalFiles: Object.keys(fileList).length,
             size: 0
         };
 
+        //check for illegal charcter in task name
+        var taskre = /sub-(.*?)_task-[a-zA-Z0-9]*[_-][a-zA-Z0-9]*(?:_acq-[a-zA-Z0-9-]*)?(?:_run-\d+)?_/g;
```
Comment (chrisgorgo) [inline validators/bids.js]: `task_re` for consistency

File: validators/bids.js [hunk 6]
```diff
@@ -97,38 +98,63 @@ var BIDS = {
...
+        var acq_re = /sub-(.*?)_task-\w+.\w+(_acq-[a-zA-Z0-9]*[_-][a-zA-Z0-9]*)(?:_run-\d+)?_/g;
+
+        for (var f in fileList) {
+            var completename = fileList[f].relativePath;
+            var acq_res = acq_re.exec(completename);
+            var res = taskre.exec(completename);
```
Comment (chrisgorgo) [inline validators/bids.js]: `task_res` for consistency

File: validators/bids.js [hunk 7]
```diff
@@ -97,38 +98,63 @@ var BIDS = {
...
+            if (res!== null){
+                self.issues.push(new Issue({
+                    file: fileList[f],
+                    code: 57,
+                    evidence: "task name contains illegal character: " + fileList[f].relativePath
+                }));
+            }
+            if (acq_res){
```
Comment (chrisgorgo) [inline validators/bids.js]: why not `!==` here as well? is there a difference?

File: validators/bids.js [hunk 8]
```diff
@@ -97,38 +98,63 @@ var BIDS = {
...
+                self.issues.push(new Issue({
+                    file: fileList[f],
+                    code: 58,
+                    evidence: "acq name contains illegal character: " + fileList[f].relativePath
+                }));
+            }
```
Comment (chrisgorgo) [inline validators/bids.js]: This is better than previous iteration, but you could make the code even more compact if you define a list (or dictionary) of triplets: regular expression, error code, evidence text prefix; and iterate over them.

File: utils/issues/list.js (hunk 4 above)
Comment (suyashdb) [reply utils/issues/list.js]: That was not intended. my branch was not up to date with master. Will get this fixed.

File: validators/bids.js (hunk 7 above)
Comment (suyashdb) [reply validators/bids.js]: its same, I will correct.

**Pull request merged**